File: LmkImageLib/Primitive/LmkBit.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;
using System.Text;
using System.Threading.Tasks;

namespace LmkImageLib
{
    /// <summary>
    /// 1-bit image
    /// </summary>
    [DataContract(Name = "")]
    public class LmkBit : ICloneable
    {
        private LmkImageClrLib.LmkBit body = new LmkImageClrLib.LmkBit();

        /// <summary>
        /// Copy constructor
        /// </summary>
        /// <param name="self">copied object</param>
        public LmkBit(LmkBit self)
            :this(self.body)
        {
        }

        /// <summary>
        /// Copy constructor
        /// </summary>
        /// <param name="bodySource"></param>
        internal LmkBit(LmkImageClrLib.LmkBit bodySource)
        {
            this.body = body.Clone();
        }

        #region ICloneable

        public LmkBit Clone()
        {
            return new LmkBit(this);
        }

        object ICloneable.Clone()
        {
            return this.Clone();
        }

        #endregion ICloneable
    }
}

File: LmkImageLib/src/LmkBit.cpp
#include "stdafx.h"
#include "LmkBit.h"

using namespace LmkImageLib;

/// <summary>
/// Constructor
/// </summary>
LmkBit::LmkBit() {
	this->data = new byte[0];
}

/// <summary>
/// Destructor
/// </summary>
LmkBit::~LmkBit() {
	// release managed resource
	delete this->data;

	this->!LmkBit();
}

/// <summary>
/// Finalizer
/// </summary>
/// <returns></returns>
LmkBit::!LmkBit() {
	// release unmanaged resource
}

/// <summary>
/// Copy constructor
/// </summary>
/// <param name="bit">original bit image</param>
/// <returns>image</returns>
LmkBit::LmkBit(LmkBit^ bit) {
	this->width = this->Width;
	this->height = this->Height;
	byte* newByte = nullptr;
	std::memcpy(newByte, this->data, sizeof(byte) * this->SizeByte);
	this->data = newByte;
	this->tags = this->tags;
}

/// <summary>
/// Copy object
/// </summary>
/// <returns>copied object</returns>
LmkBit^ LmkBit::Clone() {
	return gcnew LmkBit(this);
}

int LmkBit::Width::get() {
	return this->width;
}
int LmkBit::Height::get() {
	return this->height;
}
int LmkBit::Size::get() {
	return this->width * this->Height;
}
int LmkBit::WidthByte::get() {
	double w = (double)this->width / sizeof(byte);
	return (int)floor(w);
}
int LmkBit::HeightByte::get() {
	double h = (double)this->height / sizeof(byte);
	return (int)floor(h);
}
int LmkBit::SizeByte::get() {
	return this->WidthByte * this->HeightByte;
}
IntPtr^ LmkBit::Data::get() {
	return gcnew IntPtr(this->data);
}
String^ LmkBit::Tags::get() {
	return this->tags;
}

